File: java/app/src/main/java/com/fatchickensoftware/books/BookAdapter.java
package com.fatchickensoftware.books;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.fatchickensoftware.books.databinding.BookListItemBinding;

import java.util.ArrayList;
import java.util.List;

public class BookAdapter extends RecyclerView.Adapter<BookAdapter.BookViewHolder> {

    private LayoutInflater layoutInflater;
    List<Volume> content = new ArrayList<>();

    BookAdapter(LayoutInflater layoutInflater) {
        this.layoutInflater = layoutInflater;
    }

    @NonNull
    @Override
    public BookViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new BookViewHolder(BookListItemBinding.inflate(layoutInflater, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull BookViewHolder holder, int position) {
        VolumeInfo volumeInfo = content.get(position).volumeInfo;
        if (volumeInfo != null) {
            Context context = holder.binding.thumbnail.getContext();
            holder.binding.setVolumeInfo(volumeInfo);
            holder.binding.getRoot().setOnClickListener((View v) -> {
                context.startActivity(BookDetailActivity.createInstance(context, content.get(position)));
            });
            holder.binding.setBackgroundColor((position % 2 != 0) ? ContextCompat.getColor(context, R.color.colorRow1) : ContextCompat.getColor(context, R.color.colorRow2));
            holder.binding.setTextColor(ContextCompat.getColor(context, R.color.black));
        }
    }

    @Override
    public long getItemId(int position) {
        return content.get(position).id.hashCode();
    }

    @Override
    public int getItemCount() {
        return content.size();
    }

    static class BookViewHolder extends RecyclerView.ViewHolder {
        BookListItemBinding binding;
        BookViewHolder(BookListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: kotlin/app/src/main/java/com/fatchickensoftware/books/BookAdapter.kt
package com.fatchickensoftware.books

import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.fatchickensoftware.books.databinding.BookListItemBinding

class BookAdapter(private val layoutInflater: LayoutInflater) : RecyclerView.Adapter<BookAdapter.BookViewHolder>() {

    var content: List<Volume> = emptyList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookViewHolder =
        BookViewHolder(BookListItemBinding.inflate(layoutInflater, parent, false))

    override fun getItemId(position: Int): Long =
        content[position].id.hashCode().toLong()

    override fun onBindViewHolder(holder: BookViewHolder, position: Int) {
        content[position].volumeInfo?.let { volumeInfo ->
            val context = holder.binding.thumbnail.context

            holder.binding.volumeInfo = volumeInfo
            holder.binding.root.setOnClickListener {
                context.startActivity(BookDetailActivity.createInstance(context, content[position]))
            }
            holder.binding.backgroundColor =
                    if (position % 2 != 0) ContextCompat.getColor(context, R.color.colorRow1)
                    else ContextCompat.getColor(context, R.color.colorRow2)
            holder.binding.textColor = context?.let { ContextCompat.getColor(it, R.color.black) }
        }
    }

    override fun getItemCount() = content.size

    class BookViewHolder(internal val binding: BookListItemBinding) : RecyclerView.ViewHolder(binding.root)
}
